arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/callbacks.kt

```diff
@@ -0,0 +1,123 @@
+package arrow.fx.coroutines.stream
+
+import arrow.core.Either
+import arrow.fx.coroutines.CancelToken
+import arrow.fx.coroutines.ExitCase
+import arrow.fx.coroutines.ForkConnected
+import arrow.fx.coroutines.UnsafePromise
+import arrow.fx.coroutines.guaranteeCase
+import arrow.fx.coroutines.stream.concurrent.Queue
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.startCoroutine
+import kotlin.experimental.ExperimentalTypeInference
+
+private object END
+
+interface EmitterSyntax<A> {
+  fun onCancel(cancelF: suspend () -> Unit): Unit
+  fun emit(a: A): Unit
+  fun emit(chunk: Chunk<A>): Unit
+  fun emit(iterable: Iterable<A>): Unit
+  fun emit(vararg aas: A): Unit
+  fun end(): Unit
+}
+
+/**
+ * Creates a Stream from the given suspended block callback, allowing to emit, set cancel effects and end the emission.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.coroutines.stream.*
+ *
+ * //sampleStart
+ * suspend fun main(): Unit =
+ *   Stream.callback {
+ *       onCancel { /* cancel something */ }
+ *       emit(1)
+ *       emit(2, 3, 4)
+ *       end()
+ *     }
+ *     .compile()
+ *     .toList()
+ *     .let(::println) //[1, 2, 3, 4]
+ * //sampleEnd
+ * ```
+ *
+ * Note that if neither `end()` nor other limit operators such as `take(N)` are called,
+ * then the Stream will never end.
+ *
+ * Cancellation or errors might happen any time during emission, so it's recommended to call `onCancel` as early as
+ * possible, otherwise it's not guaranteed to be called.
+ */
+// @OptIn(ExperimentalTypeInference::class) in 1.3.70
+@UseExperimental(ExperimentalTypeInference::class)
+fun <A> Stream.Companion.callback(@BuilderInference f: suspend EmitterSyntax<A>.() -> Unit): Stream<A> =
```

Comment:
We should fix the inconsititecy between this signature and [`cancellable`](https://github.com/arrow-kt/arrow-fx/blob/e6a8618fbf2450a4084d105bf88ba563c0cc836b/arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/builders.kt#L7) & [`cancellableF`](https://github.com/arrow-kt/arrow-fx/blob/e6a8618fbf2450a4084d105bf88ba563c0cc836b/arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/builders.kt#L27) from `IO` and `Arrow Fx Coroutines`.

`suspend EmitterSyntax<A>.() -> Unit` should become `suspend EmitterSyntax<A>.() -> CancelToken`, that guarantees that when a `CancelToken` is returned it's run. When no token is returned, it can also not be run. I.e. when running blocking code inside `cancellable`. It however enforces the caller of `callback` to return a `CancelToken`.